clr/commands.py

```diff
@@ -74,44 +76,158 @@ def resolve_command(query):
+    def create(namespace):
+        return NamespaceCacheEntry(namespace.descr, namespace.longdescr, namespace.command_specs)
+
+    @property
+    def commands(self):
+        return sorted(self.command_specs.keys())
+
+class NamespaceCache:
+    """Cache introspection on command names and signatures to disk.
+
+    This allows subsequent calls to `clr help` or `clr completion` to be fast.
+    Necessary to work around the fact that many clr command namespace modules
+    import the world and initialize state on import.
+    """
+
+    def __init__(self):
+        tmpdir = os.environ.get('TMPDIR', '/tmp')
+        self.CACHE_FN = os.path.join(tmpdir, 'clr_command_cache')
+        # Clr processes are short lived. We don't close the shelve, but are
+        # careful to sync it after writes.
+        self.cache = shelve.open(self.CACHE_FN)
+
+    def get(self, namespace_key):
+        # Don't cache the system namespace. It is already loaded.
+        if namespace_key == 'system': return get_namespace('system')
+        if namespace_key not in self.cache:
+            namespace = get_namespace(namespace_key)
+            if isinstance(namespace, ErrorLoadingNamespace): return namespace
+            self.cache[namespace_key] = NamespaceCacheEntry.create(namespace)
+            self.cache.sync()
+        return self.cache[namespace_key]
 
 class System(object):
-    """System namespace for the clr tool.
+    """Namespace for system commands in the clr tool.
 
     Commands defined here will be avaliable directly without specifying a
     namespace. For example `clr help` instead of `clr system:help`. Be careful
     not to define commands here that have the same name as a defined namespace
     or it may be obscured."""
 
-    descr = 'system commands'
+    descr = 'clr built-in commands'
+
+    cache = NamespaceCache()
+
+    def cmd_clear_cache(self):
+        """Clear clr's cache.
+
+        clr caches command specs to disk to speed up help and completions.
+        Run this to clear the cache if your results are stale."""
+        # Remove file. Process exits after this, will get recreated on next run.
+        os.remove(self.cache.CACHE_FN)
+
+    def cmd_completion1(self, query=''):
+        """Completion results for first arg to clr."""
+
+        results = []
+        if ':' not in query:
+            # Suffix system commands with a space.
+            results.extend(f'{c} ' for c in self.cache.get('system').commands)
+            # Suffix namespaces with a :.
+            results.extend(f'{k}:' for k in NAMESPACE_KEYS)
+        else:
+            namespace_key, _ = query.split(':', 1)
+            results.extend(f'{namespace_key}:{c} ' for c in self.cache.get(namespace_key).commands)
+
+        print('\n'.join(r for r in results if r.startswith(query)), end='')
+
+    def cmd_profile_imports(self, *namespaces):
+        """Prints some debugging information about how long it takes to import clr namespaces."""
+        import time
```

Comment:
Is there any reason to import in a method here? I realize we don't want `clr` importing the world, but `time` is stdlib and should be super fast anyway.

Reply:
Rarely used. Was going to delete this command before the pr, but changed my mind. Moved it up because as you say it should be super fast.